ca_client.py

```diff
@@ -141,92 +165,324 @@ class CAClient(Object):
+        'legacy': '',
+        'server': 'cert_requests',
+        'client': 'client_cert_requests',
+        'application': 'application_cert_requests'}
+
+    PROCESSED_KEYS = {
+        'legacy': '',
+        'server': 'processed_requests',
+        'client': 'processed_client_requests',
+        'application': 'processed_application_requests'}
+
     def __init__(self, charm, relation_name):
         """
         :param charm: the charm object to be used as a parent object.
         :type charm: :class: `ops.charm.CharmBase`
         """
         super().__init__(charm, relation_name)
-        self._relation_name = relation_name
+        self._relation_name = self.relation_name = relation_name
         self._common_name = None
         self._sans = None
-
+        self._munged_name = self.model.unit.name.replace("/", "_")
         self._stored.set_default(ca_certificate=None, key=None,
-                                 certificate=None)
+                                 certificate=None, root_ca_chain=None)
 
         self.framework.observe(charm.on[relation_name].relation_joined,
                                self._on_relation_joined)
         self.framework.observe(charm.on[relation_name].relation_changed,
                                self._on_relation_changed)
+        self.ready_events = {
+            'legacy': self.on.tls_config_ready,
+            'server': self.on.tls_server_config_ready,
+            'client': self.on.tls_client_config_ready,
+            'application': self.on.tls_app_config_ready}
+
+    def _on_relation_joined(self, event):
+        self.on.ca_available.emit()
 
     @property
     def is_joined(self):
+        """Whether this charm has joined the relation."""
         rel = self.framework.model.get_relation(self._relation_name)
         return rel is not None
 
     @property
     def is_ready(self):
-        return all(p is not None for p in (self._stored.certificate,
-                                           self._stored.key,
-                                           self._stored.ca_certificate))
+        """Whether this charm has fulfilled the legacy certificate requests."""
+        return self._is_cert_ready('legacy')
 
-    @property
-    def _is_certificate_requested(self):
-        rel = self.framework.model.get_relation(self._relation_name)
-        if rel is None:
-            return False
-        common_name = rel.data[self.framework.model.unit].get('common_name')
-        sans = rel.data[self.framework.model.unit].get('sans')
-        if common_name is None or sans is None:
+    def _is_cert_ready(self, cert_type):
+        """Check whether there is a response for the cert_type.
+
+        :param cert_type: Certificate type
+        :type cert_type: str
+        :returns: Whether there is a response for the cert_type
+        :rtype: bool
+        """
+        try:
+            return all([
+                self.ca_certificate,
+                getattr(self._stored, cert_type, None)])
```

Comment:
I'll update the pr to explicitly set the defaults as you suggest but I don't think that allows me to remove this getattr because `cert_type` is a variable and I'm using it to dynamically construct the name of the key to check in `self._stored`